jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -936,6 +970,70 @@ protected void runJob(Runnable job)
         job.run();
     }
 
+    /**
+     * <p>Attempts to shrink the current thread from the pool if {@link #getThreads()} is greater than {@link #getMinThreads()}.</p>
+     * @return true if the current thread was shrunk, false otherwise.
+     */
+    protected boolean shrink()
+    {
+        long idleTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(getIdleTimeout());
+        while (true)
+        {
+            // No jobs available, should we shrink?
+            int threads = getThreads();
```

Comment:
How about retrying (threads-getMinThreads()) times only?

Reply:
This would just trade off a hard CPU consumption limit with a possibility of not shrinking the pool by the configured amount of threads. IMHO the current option is preferable in this case.

Reply:
On line #980,  we have a while loop with true.  Instead of true, we can have this condition. That way we can ensure total thread count doesn't go less than min thread count.

Reply:
The `while` loop is not really a loop, unless the CAS fails.
A thread would enter, and try to move the threshold. If the threshold is already in the future it exits; if it was able to move the threshold, it exits.

Reply:
`The while loop is not really a loop, unless the CAS fails.`
This I wanted to cover that when thread retries it should never go below min-threads. 

Hypothetical example:

1. min-thread = 5
2. We can shrink 10 threads in ideltimeout
3. app creates 3 more threads, total threads = 8 now
4.  All 8 threads reach to that loop at same time, and thread count may go below 5. Thus if every thread tries 8-5 = 3 times only, then only 3 thread will succeed.

Reply:
The `ensureThreads()` call in the thread's `finally` block would immediately restart any necessary thread to reach `minThread` again.

Reply:
@hiteshk25 I see what you mean now, let me try to code your suggestion.

Reply:
More over, if that happens, then we will not shrink the pool for an idleTimeout period, so will will not allow threads to die and then be recreated in a quick cycle.